# List of Contents
### 01. Importing Libraries
### 02. Importing Datasets
### 03. Data Consistency
> #### Mixed Type
> #### Missing Values
> #### Duplicates
### 04. Export Dataframe
### 05 Task

# 01. Import Libraries

In [4]:
#import libraries
import pandas as pd
import numpy as np
import os

# 02. Import Datasets

In [5]:
#path= pathway to project folder
path= r'/Users/alexmendrin/Documents/Career Foundry/Data Immersion/Achievement 4/InstaCart Basket Analysis'

In [6]:
#import original df_prods dataset
df_prods= pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'))

In [7]:
#import prepared orders_wrangled.csv data set
df_ords=pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

# 03. Data Consistency

## Mixed-Type Data

In [8]:
# create a dataframe
df_test= pd.DataFrame()

In [31]:
# create a mixed type column
df_test['mix']= ['a', 'b', 1, True]

In [32]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [33]:
#check for mixed types
for col in df_test.columns.tolist():
    weird= (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_test[weird]) > 0:
        print(col)

mix


In [34]:
#change data type to string
df_test['mix']= df_test['mix'].astype('str')

## Missing Values

In [13]:
#find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [14]:
#view missing values with subset containing only the values in question
df_nan= df_prods[df_prods['product_name'].isnull()==True]

In [15]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [16]:
df_prods.shape

(49693, 5)

In [17]:
#create new dataframe that excludes missing data
df_prods_clean= df_prods[df_prods['product_name'].isnull()==False]

In [18]:
df_prods_clean.shape

(49677, 5)

## Duplicates

In [19]:
#find duplicated rows
df_dups= df_prods_clean[df_prods_clean.duplicated()]

In [20]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [21]:
df_prods_clean.shape

(49677, 5)

In [22]:
#drop duplicates
df_prods_clean.drop_duplicates()

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [23]:
#new dataframe with no duplicates
df_prods_clean_no_dups= df_prods_clean.drop_duplicates()

In [24]:
df_prods_clean_no_dups.shape

(49672, 5)

# 04. Export Dataframe

In [26]:
#export cleaned df_prods
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

In [27]:
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

# 05. Task

### Q.2

In [28]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### It is strange that the maximum for prices are 99999, would need to look into further to understand why there is what would seem to be an outlier. Otherwise there are no other glaring irregularieties.

### Q.3

In [29]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_the_week  \
0           0   2539329        1             1                       2   
1           1   2398795        1             2                       3   
2           2    473747        1             3                       3   
3           3   2254736        1             4                       4   
4           4    431534        1             5                       4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [30]:
#checking for mixed-type data in df_ords        
for col in df_ords.columns.tolist():
    weird= (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_ords[weird]) > 0:
        print(col)

### Q.5

In [35]:
#looking for missing values in df_ords
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_the_week         0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### There are 206209 null values for days_since_prior_order, these may be due to the fact that they represent the first time that a customer fulfilled an oder, therefore they wouldn't have any previous orders

In [36]:
#observe missing values
df_nan_prior_orders= df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [37]:
df_nan_prior_orders.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_the_week  \
0            0   2539329        1             1                       2   
11          11   2168274        2             1                       2   
26          26   1374495        3             1                       1   
39          39   3343014        4             1                       6   
45          45   2717275        5             1                       3   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
11                 11                     NaN  
26                 14                     NaN  
39                 11                     NaN  
45                 12                     NaN

### Q.6

In [38]:
df_ords.shape

(3421083, 7)

In [42]:
#addressing missing data- create a new dataframe with a column for new customers
df_ords_clean= df_ords

In [43]:
#new column-new customer- indicates that there will not be an entry under days_since_prior_order
df_ords_clean['new_customer']= df_ords['days_since_prior_order'].isnull()==True

In [44]:
df_ords_clean

Unnamed: 0  order_id  user_id  order_number  orders_day_of_the_week  \
0                 0   2539329        1             1                       2   
1                 1   2398795        1             2                       3   
2                 2    473747        1             3                       3   
3                 3   2254736        1             4                       4   
4                 4    431534        1             5                       4   
...             ...       ...      ...           ...                     ...   
3421078     3421078   2266710   206209            10                       5   
3421079     3421079   1854736   206209            11                       4   
3421080     3421080    626363   206209            12                       1   
3421081     3421081   2977660   206209            13                       1   
3421082     3421082    272231   206209            14                       6   

         order_hour_of_day  days_since_prior_order  new_customer  
0                        8                     NaN          True  
1                        7                    15.0         False  
2                       12                    21.0         False  
3                        7                    29.0         False  
4                       15                    28.0         False  
...                    ...                     ...           ...  
3421078                 18                    29.0         False  
3421079                 10                    30.0         False  
3421080                 12                    18.0         False  
3421081                 12                     7.0         False  
3421082                 14                    30.0         False  

[3421083 rows x 8 columns]

#### Instead of creating a new dataframe in which I drop all of the null values, I created a new dataframe with a new column of a new customer, which indicates that they will not have an entry for days since prior order.

### Q.7

In [45]:
#check for duplicates in df_ords_clean
df_ords_dups= df_ords_clean[df_ords_clean.duplicated()]

In [46]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_the_week, order_hour_of_day, days_since_prior_order, new_customer]
Index: []

#### There were no duplicate values within the dataset, meaning that each entry represents an individual order

### Q8. 

#### Since there were no duplicates within the dataset, there is no need to get rid of any values. If there were duplicates, you would create a new dataframe and delete the duplicates using: df_ords_clean_no_dups= df_ords_dups.drop_duplicates()

### Q9.

In [47]:
#export cleaned df_ords dataset
df_ords_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

#### cleaned df_prods dataset was exported above in the exercise